[WEBSITE-654] - Move the plugin documentation hosting info from Wiki to jenkins.io (#2544)
jenkins-infra/jenkins.io: content/doc/developer/publishing/documentation.adoc

---

Documentation is an important part of any Jenkins plugin.
It includes user documentation (plugin pages, changelogs, user guidelines, etc.) and the contributor documentation (how to contribute, developer guidelines, etc.).
This section provides overview of these documentation types.

== Plugin pages

Plugin pages are hosted on the link:https://plugins.jenkins.io/[Plugin Site].
These pages are generated automatically using the metadata from the latest plugin release and an external documentation page.
External documentation can be retrieved from GitHub or from the https://wiki.jenkins.io[Jenkins Wiki].
GitHub documentation support was introduced in September 2019 (https://groups.google.com/forum/#!topic/jenkinsci-dev/VSdfVMDIW-A[announcement]).

For new plugins we recommend storing documentation in the GitHub repositories.
It allows plugin maintainers to provide the same documentation from README pages and the Jenkins plugin site,
and at the same time it allows using the Documentation-as-Code techniques when the documentation is a part of the
repository and hence all common practices can be applied:

* Pull requests and reviews
* Creating documentation in parallel with features

> * Editing docs from GitHub Web UI, with preview support

Author: It is still a feature list, not a comparison with Wiki

* Versioning, documentation for previous plugin versions can be easily accessed 

=== Using GitHub as a source of documentation

The plugin site can pull documentation from the root README pages or from other locations defined by the plugin URL (see below).
Multiple formats are supported: GitHub Markdown, Asciidoc or raw text.

To publish the plugin documentation on GitHub...

. Create a README page and put the plugin documentation there.
  This page will become a landing page for the link:https://plugins.jenkins.io/[Plugin site].
** More documentation pages can be introduced inside the repository and
linked from the README, the plugin site will display both absolute and
relative links
** Images from pages will be displayed by the plugin site as well
. Modify your project URL to point to the GitHub repository, e.g. `http://github.com/jenkinsci/your-plugin`.
  See the guidelines for Maven and Gradle below.
. Release the new plugin version.

>   After several hours, the link:https://plugins.jenkins.io/[Plugin site] with start using it as a documentation source.

Author: ```suggestion
  After several hours, the link:https://plugins.jenkins.io/[Plugin site] should display the new documentation
```

better now?

>   After several hours, the link:https://plugins.jenkins.io/[Plugin site] with start using it as a documentation source.

Reviewer: How about this?

```suggestion
  Once the link:https://plugins.jenkins.io/[plugin site] picks up the release (which can take up to a few hours) it will also display the documentation from GitHub.
```

(Assuming we invalidate plugin documentation caches when the URL changes.)

---
title: Plugin Documentation
layout: developersection
---

Documentation is an important part of any Jenkins plugin.
It includes user documentation (plugin pages, changelogs, user guidelines, etc.)

> and the contributor documentation (how to contribute, developer guidelines, etc.).

Reviewer: This part of jenkins.io generally uses line-per-sentence formatting, I would appreciate if you could keep this consistent.

This section provides overview of these documentation types.

== Plugin pages

> Plugin pages are hosted on the link:https://plugins.jenkins.io/[Plugin Site],

Reviewer: ```suggestion
Plugin pages are hosted on the link:https://plugins.jenkins.io/[Plugin Site].
```

> these pages are being generated automatically using the metadata from the latest plugin release and an external documentation page.

Reviewer: ```suggestion
These pages are generated automatically using the metadata from the latest plugin release and an external documentation page.
```

> External documentation can be retrieved from GitHub or from https://wiki.jenkins.io[Jenkins Wiki].

Reviewer: ```suggestion
External documentation can be retrieved from GitHub or from the https://wiki.jenkins.io[Jenkins Wiki].
```

Since the introduction of GitHub documentation support in September 2019
(https://groups.google.com/forum/#!topic/jenkinsci-dev/VSdfVMDIW-A[announcement]),

> this is the *recommended way* to host documentation.

Reviewer: Redundant with the following sentence.

In Jenkins we recommend storing documentation in the plugin repository.
It allows plugin maintainers to provide the same documentation from GitHub README pages and the Jenkins plugin site,
and at the same time it allows using the Documentation-as-Code techniques when the documentation is a part of the
repository and hence all common practices can be applied 

> (versioning, pull requests and reviews, creating documentation in parallel with features, editing docs from GitHub Web UI).

Reviewer: > editing docs from GitHub Web UI

It is unclear how this is a benefit of this approach. Because the GitHub UI is nicer than the wiki's?

Author: Wiki has similar feature, there is no intention to compare two approaches here

> (versioning, pull requests and reviews, creating documentation in parallel with features, editing docs from GitHub Web UI).

Reviewer: FWIW this looks like a list that deserves to be formatted as a list and extended a bit, as you probably want to highlight the benefits of this approach.

=== Using GitHub as a source of documentation

Since the introduction of GitHub documentation support in September 2019

> (https://groups.google.com/forum/#!topic/jenkinsci-dev/VSdfVMDIW-A[announcement]),

Reviewer: Already mentioned once when IMO it should be mentioned not at all.

Reviewer: Also bad formatting.

> this is the *recommended way* to host documentation. It allows plugin

Reviewer: Third time this is recommended.

maintainers to provide the same documentation from GitHub pages and the
Jenkins plugin site, and at the same time it allows using the
Documentation-as-Code techniques when the documentation is a part of the
repository and hence all common practices can be applied (versioning,
pull requests and reviews, creating documentation in parallel with

> features, editing docs from GitHub Web UI).

Reviewer: Complete duplication with previous part.

The plugin site can pull documentation from the root README pages and
from other locations. For new plugins it is recommended to use README

> pages as a source of documentation, preferably in Markdown or Asciidoc

Reviewer: Fourth time you recommend GitHub, or it's unclear what the point of this sentence is.

> format (TXT is also supported in such case).

Reviewer: Confusing.

> How to publish the plugin documentation from GitHub?

Reviewer: Looks like a header but isn't.

> * Create a README page and put the plugin documentation there

Reviewer: There is no reason this part needs to be bullet points.

> ** This page will become a landing page for https://plugins.jenkins.io/

Reviewer: Weird nesting of items looks like this was transplanted from a really bad Powerpoint presentation.

** More documentation pages can be introduced inside the repository and
linked from the README, the plugin site will display both absolute and
relative links
** Images from pages will be displayed by the plugin site as well
* Modify your project URL to point to the GitHub repository,
e.g. http://github.com/jenkinsci/your-plugin
** See the guidelines for Maven and Gradle below
* Once the new plugin version is released, link:https://plugins.jenkins.io/[Plugin site] with start using it as a documentation source

Documentation examples:

* https://plugins.jenkins.io/configuration-as-code
* https://plugins.jenkins.io/gradle
* https://plugins.jenkins.io/mailer

=== Using Jenkins Wiki as source of documentation

> Even though for new plugins it is recommended to just use GitHub README files, 

Reviewer: Fifth time.

link:https://wiki.jenkins.io[Jenkins Wiki] can still be used if there are strong reasons to do so.
See link:../wiki-page[this page] for more information.

=== Referencing the documentation page from the project file

You should link to your plugin's documentation, whether on the wiki or
elsewhere, in your plugin's pom.xml, like this:

```xml
<project>
  ...
  <url>http://github.com/jenkinsci/your-plugin</url>
  ...
</project>
```

If you're building your plugin with https://github.com/jenkinsci/gradle-jpi-plugin[Gradle], 
you can set the URL in your `+build.gradle+` like so:

```groovy
jenkinsPlugin {
  // ...
  url = 'http://github.com/jenkinsci/your-plugin'
  // ...
}
```

Or if you have a plugin written in

> https://wiki.jenkins.io/display/JENKINS/Jenkins+plugin+development+in+Ruby[Ruby],

Reviewer: Remove. Do not acknowledge this abomination exists.

you must edit your `+.pluginspec+` file like so:

```ruby
Jenkins::Plugin::Specification.new do |plugin|
  # ...
  plugin.url = 'http://github.com/jenkinsci/your-plugin'
  # ...
end
```

This ensures that the update center will list your plugin correctly once
the new plugin version is released. If this is missing, or does not
point to your Jenkins wiki page, your plugin will not be included in the
update center.

> == Adding Maintainer Information

Reviewer: Pretty sure this will be obsolete soon, so I don't think it's helpful to add this.

Maintainer information is listed for every plugin on the https://plugins.jenkins.io/[plugin site].

In your POM, make sure to include developer information, such as:

```xml
<project>
  ...
  <developers>
    <developer>

>       <id>devguy</id>

Reviewer 2: ```suggestion
      <id>exampleauthor</id>
```

>       <name>Developer Guy</name>

Reviewer 2: ```suggestion
      <name>Author Name</name>
```

>       <email>devguy@developerguy.blah</email>

Reviewer 2: Use the RFC recommended example domain
```suggestion
      <email>author@example.com</email>
```

https://github.com/jenkinsci/.github/blob/master/.github/release-drafter.adoc[Release Drafter]), 
add a link to releases page to your documentation page
(recommended)
* create a CHANGELOG file (Markdown, Asciidoc) in the repository root and link it from the documentation page
* include the changelog content in the documentation page

== Contributor documentation

For open-source plugins it is important to have contributor guidelines to attract more contributors.
GitHub offers standard ways to define guidelines and to show them to contributors,
including contributing guidelines, code of conduct, pull request templates, etc.

Some notes:

* `CONTRIBUTING` guidelines can be defined by plugin maintainers, we do not set a default guide at the moment.
  See link:https://help.github.com/en/articles/setting-guidelines-for-repository-contributors[Setting guidelines for repository contributors] for more information
* Jenkins has a link:https://jenkins.io/project/conduct/[Code of Conduct] which applies to any contributor and to any component hosted by the project.
  It is defined for all repositories using the link:https://github.com/jenkinsci/.github[jenkinsci/.github] repository,

>   plugin maintainers do not need to set it up

Reviewer: ```suggestion
  plugin maintainers do not need to set it up.
```

> * Pull request templates can be created by plugin maintainers if needed.

Reviewer: So vague as to be entirely unhelpful.

>   See link:https://help.github.com/en/articles/creating-a-pull-request-template-for-your-repository[Creating a pull request template for your repository]

Reviewer: ```suggestion
  See link:https://help.github.com/en/articles/creating-a-pull-request-template-for-your-repository[Creating a pull request template for your repository].
```

> == Useful links

Reviewer: Pretty sure the page template has metadata support for references.
